src/System.IO.Pipelines.Networking.Tls/Internal/OpenSsl/CustomBio.cs

```diff
@@ -0,0 +1,253 @@
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static class UnixLib
+        {
+            public const string CryptoDll = "libcrypto.so";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+        private static class OsxLib
+        {
+            public const string CryptoDll = "libcrypto.1.0.0.dylib";
+            [DllImport(CryptoDll, CallingConvention = CallingConvention.Cdecl)]
+            public extern static void BIO_set_flags(ref bio_st bio, BioFlags flags);
+        }
+
+        private static void BIO_set_flags(ref bio_st bio, BioFlags flags)
+        {
+            if (Interop.IsWindows)
+            {
+                WindowsLib.BIO_set_flags(ref bio, flags);
+            }
+            else if (Interop.IsOsx)
+            {
+                OsxLib.BIO_set_flags(ref bio, flags);
+            }
+            else
+            {
+                UnixLib.BIO_set_flags(ref bio, flags);
+            }
+        }
+
+        private static int FreeBio(ref bio_st bio)
+        {
+            if (bio.init == 1 && bio.shutdown == 1)
+            {
+                bio.ptr = null;
+                return 1;
+            }
+            return -1;
+        }
+
+        private static int CreateBio(ref bio_st bio)
+        {
+            bio.shutdown = 1;
+            bio.init = 1;
+            bio.num = -1;
+            return 1;
+        }
+
+        private static int WriteBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var bufferHandle = GCHandle.FromIntPtr((IntPtr)bio.ptr);
+            var bufferPipeline = (IPipelineWriter)bufferHandle.Target;
+            int numberOfBytesRemaing = numberOfBytes;
+            var buffer = bufferPipeline.Alloc();
+            while (numberOfBytesRemaing > 0)
+            {
+                int sizeToWrite = Math.Min(MaxBlockSize, numberOfBytesRemaing);
+                buffer.Ensure(sizeToWrite);
+                var writeSpan = new Span<byte>(buff, sizeToWrite);
+                writeSpan.CopyTo(buffer.Memory.Span);
+                buffer.Advance(sizeToWrite);
+                numberOfBytesRemaing -= sizeToWrite;
+            }
+            buffer.Commit();
+            bio.num_write = (uint)numberOfBytes;
+            BIO_set_flags(ref bio, BioFlags.BIO_FLAGS_WRITE | BioFlags.BIO_FLAGS_SHOULD_RETRY);
+            return numberOfBytes;
+        }
+
+        public static int NumberOfWrittenBytes(InteropBio.BioHandle handle)
+        {
+            var bio = (bio_st*)handle.Handle;
+            var bytes = (int)bio[0].num_write;
+            bio[0].num_write = 0;
+            return bytes;
+        }
+
+        private static int ReadBio(ref bio_st bio, void* buff, int numberOfBytes)
+        {
+            var buffer = Unsafe.Read<ReadableBuffer>(bio.ptr);
```

Comment:
I hadn't looked closely enough at ReadableBuffer before the first comment. Now I have slightly different concerns. I still think length checking is important (yes it's against a length you stored earlier, but it prevents future bugs). More importantly, ReadableBuffer contains GC pointers. It's not safe to copy its contents to native memory and back as that can introduce [GC holes](https://github.com/dotnet/coreclr/blob/master/Documentation/coding-guidelines/clr-code-guide.md#2.1.1). If you're just trying to round-trip it, I'd suggest using a GCHandle instead.